synapseclient/client.py

```diff
@@ -1770,82 +1773,146 @@ def returnDict(destination):
                      'files': [None] if destination is None else [os.path.basename(destination)],
                      'cacheDir': None if destination is None else os.path.dirname(destination) }
 
-        if expected_md5 == '':
-            expected_md5 = None
+        return returnDict(self._download_with_retries(url, destination, file_handle_id, expected_md5))
```

Comment:
I broke it up into 2 functions for clarity - separating the retry logic from the loop handling redirects.

I also like that __download_with_retries_ keeps a copy of the original URL in its scope leaving __download_ free to modify its copy or URL as it gets redirected. Otherwise, you'd have to stash a copy of the original so you could refresh the presigned and time-limited S3 URL.